cub/device/dispatch/dispatch_batch_memcpy.cuh

```diff
@@ -0,0 +1,670 @@
+ * buffers to their corresponding destination buffer.
+ */
+
+#pragma once
+
+/// Optional outer namespace(s)
+#include <cstdint>
+
+#include <cub/agent/agent_batch_memcpy.cuh>
+#include <cub/agent/single_pass_scan_operators.cuh>
+#include <cub/config.cuh>
+#include <cub/detail/temporary_storage.cuh>
+#include <cub/thread/thread_search.cuh>
+
+#include <thrust/system/cuda/detail/core/triple_chevron_launch.h>
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+
+/**
+ * Parameterizable tuning policy type for AgentBatchMemcpy
+ */
+template <uint32_t _BLOCK_THREADS, uint32_t _BYTES_PER_THREAD>
+struct AgentBatchMemcpyLargeBuffersPolicy
+{
+  enum
+  {
+    /// Threads per thread block
+    BLOCK_THREADS = _BLOCK_THREADS,
+    /// The number of bytes each thread copies
+    BYTES_PER_THREAD = _BYTES_PER_THREAD
+  };
+};
+
+/**
+ * Initialization kernel for tile status initialization (multi-block)
+ */
+template <typename BufferOffsetScanTileStateT, typename BlockOffsetScanTileStateT, typename TileOffsetT>
+__global__ void InitTileStateKernel(BufferOffsetScanTileStateT buffer_offset_scan_tile_state,
+                                    BlockOffsetScanTileStateT block_offset_scan_tile_state,
+                                    TileOffsetT num_tiles)
+{
+  // Initialize tile status
+  buffer_offset_scan_tile_state.InitializeStatus(num_tiles);
+  block_offset_scan_tile_state.InitializeStatus(num_tiles);
+}
+
+/**
+ * Kernel that copies buffers that need to be copied by at least one (and potentially many) thread blocks.
+ */
+template <typename ChainedPolicyT,
+          typename BufferOffsetT,
+          typename InputBufferIt,
+          typename OutputBufferIt,
+          typename BufferSizeIteratorT,
+          typename BufferTileOffsetItT,
+          typename TileT,
+          typename TileOffsetT>
+__launch_bounds__(int(ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT::BLOCK_THREADS)) __global__
+  void MultiBlockBatchMemcpyKernel(InputBufferIt input_buffer_it,
+                                   OutputBufferIt output_buffer_it,
+                                   BufferSizeIteratorT buffer_sizes,
+                                   BufferTileOffsetItT buffer_tile_offsets,
+                                   TileT buffer_offset_tile,
+                                   TileOffsetT last_tile_offset)
+{
+  using StatusWord    = typename TileT::StatusWord;
+  using ActivePolicyT = typename ChainedPolicyT::ActivePolicy::AgentLargeBufferPolicyT;
+
+  constexpr uint32_t BLOCK_THREADS    = ActivePolicyT::BLOCK_THREADS;
+  constexpr uint32_t ITEMS_PER_THREAD = ActivePolicyT::BYTES_PER_THREAD;
+  constexpr uint32_t ITEMS_PER_BLOCK  = BLOCK_THREADS * ITEMS_PER_THREAD;
+
+  BufferOffsetT num_blev_buffers = buffer_offset_tile.LoadValid(last_tile_offset);
+
+  uint32_t tile_id = blockIdx.x;
+
+  // Make sure we've fetched read the total number of block-level buffers into num_blev_buffers
+  __syncthreads();
```

Comment:
`num_blev_buffers` is just a local variable now. Should we remove the sync along with the comment?

Reply:
Totally! I've removed it. Thanks for catching this 👍